sparse-checkout: don't update sparse-checkout file on error

derrickstolee opened the pull request:
If the index has an unstaged change in a file that will be dropped
by a change to the sparse-checkout file, then the index update
will fail. If this happens during 'git sparse-checkout set', then
the entire command will fail. However, the sparse-checkout file is
still updated, so the working directory is in a half-finished state.

Update the sparse-checkout builtin to run `unpack_trees()`
directly instead of through an external `git read-tree` process.
While doing so, allow passing a set of patterns, so we can
validate the patterns before trying to write to the sparse-checkout
file.

Also, write sparse-checkout under a .lock and take the
sparse-checkout lock around the unpack_trees() check.

Resolves #198.

kewillford commented on builtin/sparse-checkout.c:
```diff
@@ -259,9 +259,21 @@ static void write_cone_to_file(FILE *fp, struct pattern_list *pl)
 
 	sparse_filename = get_sparse_checkout_filename();
+	strbuf_addstr(&backup_filename, sparse_filename);
+	strbuf_addstr(&backup_filename, ".backup");
```
Maybe this is overkill but should we be using the `.lock` mechanisms in git to write the change to the `.lock` file and delete that on failure or rename on success?

kewillford replied:
Also thinking about this a little and do we need the protection of the `.lock` to prevent 2 `sparse-checkout` commands from running at the same time?

derrickstolee replied:
You are probably right about the `.lock` file, but that won't solve the situation here: we need to commit the `.lock` file before the update can work on the new data.

However, I do wonder if we can do something more clever: we have a `struct pattern_list`. Can we try running `clear_ce_flags()` based on this list and get an error from that instead? I will investigate.

derrickstolee committed:
sparse-checkout: update working directory in-process

Signed-off-by: Derrick Stolee <dstolee@microsoft.com>

derrickstolee committed:
sparse-checkout: back out a change with conflicts

Signed-off-by: Derrick Stolee <dstolee@microsoft.com>

derrickstolee committed:
sparse-checkout: allow cone-type patterns for full tree

Due to an oversight when handling cone patterns in sparse-checkout,
the full cone patternset (only "/*") would not fill any entries in
the hashsets, and only files at root would be populated.

For cone mode patterns, the first two patterns are _usually_ the
following:

	/*
	!/*/

However, we instead need to handle the case where that second
line does not exist. Due to the MUSTBEDIR and NEGATIVE flags, the
second pattern becomes "/*".

Create a new option in struct pattern_list to store this "full
cone" mode. When we see the "/*" pattern with no flags, we toggle
full_cone to on. If we later see "/*" pattern with MUSTBEDIR and
NEGATIVE flags, then turn full_cone off.

Signed-off-by: Derrick Stolee <dstolee@microsoft.com>

derrickstolee committed:
fixup! sparse-checkout: init and set in cone mode

Signed-off-by: Derrick Stolee <dstolee@microsoft.com>

derrickstolee commented on builtin/sparse-checkout.c:
```diff
@@ -285,8 +339,8 @@ static void strbuf_to_cone_pattern(struct strbuf *line, struct pattern_list *pl)
+	if (line->buf[0] != '/')
+		strbuf_insert(line, 0, "/", 1);
 
 	if (!line->len)
```
remove this now that it is useless.

derrickstolee commented on builtin/sparse-checkout.c:
```diff
@@ -343,11 +399,11 @@ static int sparse_checkout_set(int argc, const char **argv, const char *prefix)
 		} else {
 			for (i = 0; i < argc; i++)
-				add_pattern(argv[i], argv[i], strlen(argv[i]), &pl, 0);
+				add_pattern(argv[i], empty_base,0, &pl, 0);
```
nit: whitespace

wilbaker commented on builtin/sparse-checkout.c:
```diff
@@ -60,18 +65,53 @@ static int sparse_checkout_list(int argc, const char **argv)
-	if (run_command_v_opt(argv.argv, RUN_GIT_CMD)) {
-		error(_("failed to update index with new sparse-checkout paths"));
-		result = 1;
+	if (repo_read_index_unmerged(r))
```
> If the index has an unstaged change in a file that will be dropped
by a change to the sparse-checkout file, then the index update
will fail. 

For my own understanding, what happens to untracked files that are outside of the new cone?  Are they left in place or deleted?

derrickstolee replied:
From my testing, it looks like untracked/non-ignored files remain. Everything else in the parent directory will be deleted.

wilbaker commented on dir.h:
```diff
@@ -71,6 +71,7 @@ struct pattern_list {
 	 * excludes array above. If non-zero, that check succeeded.
 	 */
 	unsigned use_cone_patterns;
+	unsigned full_cone;
```
>However, we instead need to handle the case where that second
line does not exist. Due to the MUSTBEDIR and NEGATIVE flags, the
second pattern becomes "/*".
>
>Create a new option in struct pattern_list to store this "full
cone" mode. When we see the "/*" pattern with no flags, we toggle
full_cone to on. If we later see "/*" pattern with MUSTBEDIR and
NEGATIVE flags, then turn full_cone off.

For my own understanding, are these flags set when the file is parsed?

derrickstolee replied:
Yes, this is part of the `parse_path_pattern()` method in `dir.c`, which is called by `add_pattern()` inside `add_patterns_from_buffer()`.

derrickstolee replied:
That method also strips out the tail `/` for directories (`MUSTBEDIR` flag) and the preceding `!` for negations (`NEGATIVE` flag).

wilbaker commented on dir.c:
```diff
@@ -622,8 +622,17 @@ static void add_pattern_to_hashsets(struct pattern_list *pl, struct path_pattern
 
-	if (!strcmp(given->pattern, "/*"))
+	if (given->flags & PATTERN_FLAG_NEGATIVE &&
+	    given->flags & PATTERN_FLAG_MUSTBEDIR &&
```
Under what circumstances would only one of `PATTERN_FLAG_NEGATIVE` and `PATTERN_FLAG_MUSTBEDIR` be set?

derrickstolee replied:
```
| Pattern        | Flags              |
|----------------+--------------------|
| /A/B/f.txt     | (none)             |
|----------------+--------------------|
| !/A/B/f.txt    | NEGATIVE           |
|----------------+--------------------|
| /A/B/          | MUSTBEDIR          |
|----------------+--------------------|
| !/A/B/         | MUSTBEDIR|NEGATIVE |
|----------------+--------------------|
```

wilbaker reviewed (commented):
I left a few questions to help with my understanding of this code.

derrickstolee committed:
sparse-checkout: write using lockfile

If two 'git sparse-checkout set' subcommands are launched at the
same time, the behavior can be unexpected as they compete to write
the sparse-checkout file and update the working directory.

Take a lockfile around the writes to the sparse-checkout file. In
addition, acquire this lock around the working directory update
to avoid two commands updating the working directory in different
ways.

Signed-off-by: Derrick Stolee <dstolee@microsoft.com>

derrickstolee commented:
Ping on this one. I adapted these commits into [v3 of the sparse-checkout series](https://public-inbox.org/git/pull.316.v3.git.gitgitgadget@gmail.com/T/#u) which went upstream yesterday. I expect to need a re-roll due to clashes with some hashmap changes.

cc: @kewillford @wilbaker

kewillford commented on dir.c:
```diff
@@ -1285,6 +1294,9 @@ enum pattern_match_result path_matches_pattern_list(
 	}
 
+	if (pl->full_cone)
+		return MATCHED;
```
Why is the `MATCHED` and not `MATCHED_RECURSIVE`?

derrickstolee replied:
Unfortunately, `MATCHED_RECURSIVE` doesn't work if we return it for files.

wilbaker commented on dir.c:
```diff
@@ -622,8 +622,17 @@ static void add_pattern_to_hashsets(struct pattern_list *pl, struct path_pattern
+	if (given->flags & PATTERN_FLAG_NEGATIVE &&
+	    given->flags & PATTERN_FLAG_MUSTBEDIR &&
+	    !strcmp(given->pattern, "/*")) {
+		pl->full_cone = 0;
```
Apologies if we already discussed this (I'm having trouble finding it if we did), but is there somewhere that git enforces that the sparse-checkout file is "valid", or are the entries guaranteed to be in a particular order?

I ask because I'm wondering if after setting `pl->full_cone = 0` here we might read an entry that sets it to 1 below (or vice-versa).  Is that the expected behavior?

derrickstolee replied:
The order is absolutely important in the `sparse-checkout` file. In our typical cone-based patterns, we start with these two lines:

```
/*     # include all files
!/*/   # except those in folders
```

The order gives the meaning to the negation. The last pattern to match the file `/A/file.txt` is `!/*/`, so the sparse-checkout says "do not include this file". (If no pattern matches at all, the default is "do not include this file".)

If they are reversed:

```
!/*/     # do not include things in folders
/*       # include everything
```

then the sparse-checkout file will match everything, as the _last_ pattern to match is a positive pattern.

So, this code is saying "If I see `!/*/`, then set `full_cone` to zero. If I see `/*`, then set `full_cone` to true."

If there are other patterns that make this file not look like a cone-based file, then `use_cone_patterns` is set to zero and `full_cone` is ignored.

kewillford commented on dir.c:
```diff
@@ -622,8 +622,17 @@ static void add_pattern_to_hashsets(struct pattern_list *pl, struct path_pattern
+	}
+
+	if (!given->flags && !strcmp(given->pattern, "/*")) {
+		pl->full_cone = 1;
```
I'm a little confused on what `full_cone` means.  Will you explain the intent of the flag?

derrickstolee replied:
The point is to say "all files will match". The place this comes in is `git sparse-checkout disable` with cone mode enabled. It needs to set the sparse-checkout file to "/*\n" and then update the working directory before removing the config. Without this fix, the cone-mode patterns would not work.

kewillford commented on builtin/sparse-checkout.c:
```diff
@@ -60,18 +65,53 @@ static int sparse_checkout_list(int argc, const char **argv)
+	o.head_idx = -1;
+	o.src_index = r->index;
+	o.dst_index = r->index;
+	o.skip_sparse_checkout = 0;
```
How did you decide these were the right options for unpack_trees?

derrickstolee replied:
I read the code in `builtin/read-tree.c` and tracked what would happen in a call to `git read-tree -mu HEAD`.

kewillford commented on builtin/sparse-checkout.c:
```diff
@@ -343,11 +405,11 @@ static int sparse_checkout_set(int argc, const char **argv, const char *prefix)
 				size_t len;
 				char *buf = strbuf_detach(&line, &len);
-				add_pattern(buf, buf, len, &pl, 0);
+				add_pattern(buf, empty_base, 0, &pl, 0);
```
Will you explain this change for me?

derrickstolee replied:
Before, we were not using the `pattern_list` for anything other than tracking the patterns to set in the sparse-checkout file.

Now, we want to use the `pattern_list` in-memory to update the working directory.

The previous way I was calling `add_pattern()` was wrong and led to errors when using these patterns in-memory.

kewillford commented on builtin/sparse-checkout.c:
```diff
@@ -252,28 +298,45 @@ static void write_cone_to_file(FILE *fp, struct pattern_list *pl)
+		rollback_lock_file(&lk);
+		free(sparse_filename);
+		clear_pattern_list(pl);
+		update_working_directory(NULL);
```
Wouldn't we need to load the previous patterns and pass that to `update_working_directory`?

derrickstolee replied:
By using `NULL`, it will load the patterns from the `sparse-checkout` file in the normal way.

kewillford reviewed (commented):
Questions to help me better understand.

wilbaker reviewed (approved):
Looks good to me, but @kewillford is more of an expert on this code so you might want to wait for his approval before merging.

kewillford commented on builtin/sparse-checkout.c:
```diff
@@ -60,18 +65,53 @@ static int sparse_checkout_list(int argc, const char **argv)
+
+	cache_tree_free(&r->index->cache_tree);
+
+	repo_hold_locked_index(r, &lock_file, LOCK_DIE_ON_ERROR);
```
I'm wondering if we should hold the `index.lock` for the entire `sparse-checkout` command to prevent races?  For example running a checkout after this code commits the `index.lock` but it runs before the sparse-checkout file was updated? 

Since commands get the index.lock before modifying the index, I'm thinking the sequence should be:
1. Get the `index.lock`
2. Get the `sparse-checkout.lock`
3. Update the working directory and sparse-checkout (I don't think order matters here since we have both locks)
4. Commit the `sparse-checkout`
5. Commit the `index`

derrickstolee replied:
I'll give that a try, thanks!

derrickstolee replied:
I rearranged the code to do exactly this in 5cd306f. Thanks!

kewillford reviewed (approved):
Looks good. Approved with the one suggestion about locking.

derrickstolee committed:
sparse-checkout: update locking pattern

There is a narrow window for race conditions while one process
updates the sparse-checkout file and another process tries to
do a checkout.

To avoid these races, follow this pattern in 'git sparse-checkout
set':

  1. Take index.lock.
  2. Take sparse-checkout.lock.
  3. Update the working directory according to in-memory patterns.
  4. Write patterns to sparse-checkout.lock.
  5. Commit sparse-checkout.lock.
  6. Commit index.lock.

This avoids the race condition, as any checkout must first take
the index.lock.

The code is organized into a new "struct update_data" so we can
hold the multiple structs and options across this time period.

Helped-by: Kevin Willford <kewillford@microsoft.com>
Signed-off-by: Derrick Stolee <dstolee@microsoft.com>

kewillford commented on builtin/sparse-checkout.c:
```diff
@@ -252,28 +327,53 @@ static void write_cone_to_file(FILE *fp, struct pattern_list *pl)
-	free(sparse_filename);
+	fflush(fp);
 
+	/* commit index.lock file */
+	finish_update_data(ud);
+	free(ud);
```
Should the commit of the `index.lock` be the last thing to happen so after the commit of the `sparse-checkout.lock`?  I'm not sure what git reads first the sparse-checkout file or the index.  Not sure it matters too much either since both renames should happen rather fast. But since the commit has
```
  5. Commit sparse-checkout.lock.
  6. Commit index.lock.
```
Thought I would ask.

derrickstolee replied:
Since the `sparse-checkout` builtin is the only thing looking at `sparse-checkout.lock`, this won't affect parallel checkouts. A `checkout`-like command would take `index.lock` before reading the `sparse-checkout` file.

To avoid deadlocks between competing `git sparse-checkout set` commands, it is best to "pop" the locks in a stack.

derrickstolee replied:
And yes, I definitely did them in the wrong order in the code. Thanks for catching that!
